### ISSUE TITLE
Reestablish DTLS connection when not using status-server messages - 2024

### ISSUE THREAD

**Issue by fue36** [radsecproxy/radsecproxy#148]
I'm having a identical issue as described here.  https://github.com/radsecproxy/radsecproxy/issues/61

DTLS server (Cisco ISE) does not support Status-Server packets.  After ISE DTLS Idle timeout, which is configurable between 60-600 seconds, it will no longer respond to requests until the DTLS is reestablished.  radsecproxy will not attempt to reestablish connection unless StatusServer is used within the server directive, regardless of how many failed responses there are.

I have been playing with the server block setting RetryInterval and RetryCount, currently I have them set to 3 & 10, trying to get to authenticate consistently after the server timed out the connection with StatusServer minimal.  But the only consistently so far as lack of there of.  Logins are extremely slow, taking from 15 secs to the occasional timeout.  Setting Status-Server On, mostly allows authentication, but it's ugly packet mess, since ISE does not support.  Only other thing I can think of is to restart the radsecproxy after _x_ amount of minutes and setting DTLS timeout on the ISE server to maximum.

Maybe having another setting within the server block for the use case?  

Fictional setting for server section (DTLSReconnect) - Always re-establish DTLS connection before inserting packets in the server queue after _x_ amount of seconds after the last server response.

```
Server {
  DTLSReconnect 50
}

```

**Comment by fue36** [comment]
Looking through packet captures, when the ISE server closes the connection, it sends a DTLS 1.2 packet that wireshark decodes as "Alert (Fatal, Description: Close Notify)".  It seems to that the root of the issue is radsecproxy not interpreting the 'close notify' message and just assuming the connection is always open.  radsecproxy should know that the connection closed and reestablish connection before transmitting the radius message.

**Comment by fmauchle** [comment]
Thanks for the detailed report.
I'm trying to reproduce this issue, but without success so far. When testing between two tradsecproxy (which does similar connection timeout on the server side), when the server closes the connetion, it sends a TLS alert close notify, which the client correctly reads and re-opens the connection.

It's actually OpenSSL which handles the TLS close notify, which simply informs the client code about the fact. But it's all there. Which version of OpenSSL are you using?

Another curious thing: you state that wireshark was able to decode the close notify. These should normally be encrypted. Are you using TLS decryption or was the close notify actually sent in clear?

Trying to spot the difference...

**Comment by fmauchle** [comment]
I agree the certificate should only matter at connection establishment.
However, the ICMP Port unreachable is something to investigate closer. This normally indicates that the client has already closed its UDP socket (the ICMP is generated by the kernel network stack). One thing to check is if the TLS close notify is actually sent to the correct port (of the DTLS connectio to be closed - just to avoid confusion with overlapping connections and late replies).
Another way this can often happen is some firewalls in between (or an iptables or nftables on the host itself) that do their own connection tracking and apply session timeouts that are shorter than on the server itself.

**Comment by fue36** [comment]
The problem is firewalld on RHEL / Alma, which runs by default.  When I stop the firewalld service, the kernel network stack does not generate the ICMP Port unreachable. With 'StatusServer minimal' in the radsecproxy config, as soon as the server closes the connection with 'TLS Close Notify' the radsecproxy immediately logs 'sslreadtimeout: connection lost: EOF' and the connection is re-opened and authentication succeeds as expected.

With firewalld enabled, when the server sends the 'TLS close notify' message, the client responds with 'ICMP Unreachable'.   radsecproxy does not react and continues to send DTLS authentication datagram on same src / dst pairs.  The server ignores the requests and does not respond.

Furthermore firewalld does not log anything in /var/log/firewalld when configured with 'LogDenied=all'.  I will try to come up with a better solution than to completely disable firewalld.

**Comment by fue36** [comment]
Ok, so Redhat support stated they only support simple rules for firewalld using nftables.  Inserting custom rules into firewalld using nftables is not supported.  So I disabled the firewall completely, no biggie for me.

However I am still having issues.  Now what happens is occasionally after the DTLS server sends 'TLS close notify' message, radsecproxy doesn't acknowledge it and thinks the connection is open (no '_sslreadtimeout: connection lost: EOF_' is logged).  Sometimes it works for a day or so before it loses sync, other times its quicker.  Now, since the firewall is disabled, there is no ICMP unreachable message from the client.  I can't figure a rhyme or reason.  What I do know, through packet captures, is that the client does always receive the 'TLS close notify' message from the server.

For now I am working around this for lack of better solution by setting the DTLS server timeout to 600 seconds and restarting radsecproxy on the client every 10 minutes through cron.
